src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DefaultVisualStudioDocumentTrackerSubscriber.cs

```diff
@@ -0,0 +1,304 @@
+                {
+                    TryAddCSharpBuffer(csharpBuffer);
+                }
+                _bufferGraphChangeListener = new BufferGraphChangeListener(bufferGraph, OnBufferGraphChange);
+
+                _onCSharpBuffersChanged(_documentTracker, _csharpBuffers);
+            }
+
+            private void TryAddCSharpBuffer(ITextBuffer textBuffer)
+            {
+                if (!IsCSharpBuffer(textBuffer))
+                {
+                    return;
+                }
+
+                AttachWorkspaceChangeListener(textBuffer, OnWorkspaceChangedForBuffer);
+
+                _csharpBuffers.Add(textBuffer);
+            }
+
+            private void TryRemoveCSharpBuffer(ITextBuffer textBuffer)
+            {
+                if (textBuffer.Properties.TryGetProperty<WorkspaceChangeListener>(typeof(WorkspaceChangeListener), out var changeListener))
+                {
+                    textBuffer.Properties.RemoveProperty(typeof(WorkspaceChangeListener));
+                    changeListener.Dispose();
+                    _csharpBuffers.Remove(textBuffer);
+                }
+            }
+
+            private void OnBufferGraphChange(GraphBuffersChangedEventArgs args)
+            {
+                foreach (var addedBuffer in args.AddedBuffers)
+                {
+                    TryAddCSharpBuffer(addedBuffer);
+                }
+
+                foreach (var removedBuffer in args.RemovedBuffers)
+                {
+                    TryRemoveCSharpBuffer(removedBuffer);
+                }
+
+                _onCSharpBuffersChanged(_documentTracker, _csharpBuffers);
+            }
+
+            private void OnWorkspaceChangedForBuffer(ITextBuffer textBuffer)
+            {
+                if (!IsCSharpBuffer(textBuffer))
+                {
+                    return;
+                }
+
+                _onCSharpBuffersChanged(_documentTracker, _csharpBuffers);
+            }
+
+            private void AttachWorkspaceChangeListener(
+                ITextBuffer textBuffer,
+                Action<ITextBuffer> onWorkspaceChanged)
+            {
+                Debug.Assert(!textBuffer.Properties.ContainsProperty(typeof(WorkspaceChangeListener)));
+
+                var workspaceChangeListener = new WorkspaceChangeListener(textBuffer, onWorkspaceChanged);
+                textBuffer.Properties[typeof(WorkspaceChangeListener)] = workspaceChangeListener;
+            }
+
+            public void Dispose()
+            {
+                _bufferGraphChangeListener.Dispose();
+
+                for (var i = _csharpBuffers.Count - 1; i >= 0; i--)
+                {
+                    TryRemoveCSharpBuffer(_csharpBuffers[i]);
+                }
+            }
+
+            private static bool IsCSharpBuffer(ITextBuffer textBuffer) =>
+                textBuffer.ContentType.IsOfType(CSharpContentType) || textBuffer.ContentType.IsOfType(LiveShareCSharpContentType);
```

Comment:
Does the C# LSP content type not inherit from the C# content type? That seems a little wierd.

Reply:
It doesn't, LiveShare changes the content type of the Buffer from CSharp -> C#_LSP, and re-apply all of the functionality for whatever reason. Don't really know the whole background as to why they do that but there's a race where they rename the buffer before we see it so we need to also consider it as the C# buffer 😢